### PR TITLE
Use uint64 offset, Redis Streams support

### PR DESCRIPTION
**Pull request by FZambia**
Massive library refactoring. Relates to #96 

Things introduced:
1) Offset `uint64` field instead of Seq and Gen `uint32`. With backwards compatibility in mind via global flags that will be removed in v1 release
2) Redis Streams support
3) Slightly refactored Engine interface and Publish API.

Redis streams performance compares to current Redis lists approach but much more effective in terms of recover inside large streams.

For example for recovering last 5 messages in history with different sizes:

100 messages in stream:

```
lists-4         6644	    179385 ns/op	   26155 B/op	     820 allocs/op
streams-4       7137	    168041 ns/op	    3168 B/op	     111 allocs/op
```

1000 messages in stream:

```
lists-4         1516	    751995 ns/op	  249733 B/op	    8021 allocs/op
streams-4       6680	    167779 ns/op	    3171 B/op	     111 allocs/op
```

10000 messages in stream:

```
lists-4         140	   8345434 ns/op	 2535549 B/op	   80025 allocs/op
streams-4       6343	    166731 ns/op	    3181 B/op	     111 allocs/op
```  

As we can see Redis streams allow us to maintain the same performance to recover part of stream.

TODO:
- support offset in libraries that already has recovery feature implemented: `centrifuge-js` and `centrifuge-go`
- standardize corner cases stream behaviours

### CODE REVIEW

File: internal/recovery/helpers.go [hunk 1]
```diff
@@ -26,43 +22,36 @@ func uniquePublications(s []*protocol.Publication) []*protocol.Publication {
...
 // MergePublications ...
-func MergePublications(recoveredPubs []*protocol.Publication, bufferedPubs []*protocol.Publication) ([]*protocol.Publication, bool) {
+func MergePublications(recoveredPubs []*protocol.Publication, bufferedPubs []*protocol.Publication, isLegacyOrder bool) ([]*protocol.Publication, bool) {
 	if len(bufferedPubs) > 0 {
 		recoveredPubs = append(recoveredPubs, bufferedPubs...)
 	}
 	sort.Slice(recoveredPubs, func(i, j int) bool {
-		if recoveredPubs[i].Gen != recoveredPubs[j].Gen {
-			return recoveredPubs[i].Gen > recoveredPubs[j].Gen
+		if isLegacyOrder {
```
Comment (cristaloleg) [inline internal/recovery/helpers.go:30]: We can move this `if` out of sort. perf++, readability++

File: engine_redis.go [hunk 2]
```diff
@@ -1674,32 +1858,97 @@ func mapStringClientInfo(result interface{}, err error) (map[string]*ClientInfo,
...
+		if push.Type != protocol.PushTypePublication {
+			return nil, fmt.Errorf("wrong message type in history: %d", push.Type)
+		}
+
+		var pub protocol.Publication
+		err = pub.Unmarshal(push.Data)
+		if err != nil {
+			return nil, fmt.Errorf("can not unmarshal value to Pub: %v", err)
+		}
+		if !useSeqGen {
```
Comment (cristaloleg) [inline engine_redis.go:1926]: invert?

File: engine_redis.go [hunk 3]
```diff
@@ -1711,16 +1960,18 @@ func sliceOfPubs(result interface{}, err error) ([]*Publication, error) {
...
 		err = pub.Unmarshal(push.Data)
 		if err != nil {
 			return nil, fmt.Errorf("can not unmarshal value to Pub: %v", err)
 		}
 
-		pub.Seq = seq
-		pub.Gen = gen
-		pubs[j] = &pub
-		j++
+		if !useSeqGen {
```
Comment (cristaloleg) [inline engine_redis.go:1969]: invert?

Comment (FZambia) [reply engine_redis.go]: fixed

File: engine_redis.go (hunk 2 above)
Comment (FZambia) [reply engine_redis.go]: fixed

File: internal/recovery/helpers.go (hunk 1 above)
Comment (FZambia) [reply internal/recovery/helpers.go]: fixed (if I understood this point right)

**Pull request merged**